webapp/src/ts/services/indexed-db.service.ts

```diff
@@ -0,0 +1,99 @@
+import { Inject, Injectable } from '@angular/core';
+import { DOCUMENT } from '@angular/common';
+
+import { DbService } from '@mm-services/db.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class IndexedDbService {
+  private readonly LOCAL_DOC = '_local/indexeddb-placeholder';
+  private loadingLocalDoc: Promise<IndexedDBDoc> | undefined;
+  private hasDatabasesFn: boolean;
+  private isUpdatingLocalDoc: boolean;
+
+  constructor(
+    private dbService: DbService,
+    @Inject(DOCUMENT) private document: Document
+  ) {
+    // Firefox doesn't support the databases() function. Issue: https://github.com/medic/cht-core/issues/8777
+    this.hasDatabasesFn = !!this.document.defaultView?.indexedDB?.databases;
+  }
+
+  async getDatabaseNames() {
+    if (this.hasDatabasesFn) {
+      const databases = await this.document.defaultView!.indexedDB.databases();
```

Comment:
The check `if (this.hasDatabasesFn)` makes it reasonable to assume that indexedDB.databases is available hence it is safe to use the ! operator.  But from a defensive programming view, perhaps using `const databases = await this.document.defaultView?.indexedDB?.databases();` would be more robust.

Reply:
yes, I was back and forth on this. I applied the suggestion